pandas/_libs/tslibs/np_datetime.pyx

```diff
@@ -450,3 +451,43 @@ cdef int op_to_op_code(op):
         return Py_GE
     if op is operator.gt:
         return Py_GT
+
+
+@cython.overflowcheck(True)
+cdef int64_t get_conversion_factor(NPY_DATETIMEUNIT from_unit, NPY_DATETIMEUNIT to_unit) except? -1:
+    """
+    Find the factor by which we need to multiply to convert from from_unit to to_unit.
+    """
+    if (
+        from_unit == NPY_DATETIMEUNIT.NPY_FR_GENERIC
+        or to_unit == NPY_DATETIMEUNIT.NPY_FR_GENERIC
+    ):
+        raise ValueError("unit-less resolutions are not supported")
+    if from_unit > to_unit:
+        raise ValueError
+
+    if from_unit == to_unit:
+        return 1
+
+    if from_unit == NPY_DATETIMEUNIT.NPY_FR_W:
```

Comment:
Might be worth exploring a iterative solution in the future for perf maybe?

e.g.

```
factor = 1
while from_unit != to_unit:
    factor *= factor_map[from_unit] e.g. 7 (day to week)
    from_unit = factor_map_step[from_unit] e.g. `NPY_DATETIMEUNIT.NPY_FR_W`
return factor
```

Reply:
Gotcha. This current version is fairly clear already so IMO what's here in this PR is probably just worth keeping for the clarity